master/buildbot/process/buildstep.py

```diff
@@ -421,8 +437,9 @@ def getResultSummary(self):
         return {'step': stepsumm}
 
     @defer.inlineCallbacks
-    def getBuildResultSummary(self):
-        summary = yield self.getResultSummary()
+    def getBuildResultSummary(self) -> Generator[defer.Deferred[Any], Any, dict[str, str]]:
```

Comment:
This is a bit hard to understand, especially to new contributors to Buildbot.

I wonder if a better path would be to convert the function to async and decorate it using `async_to_deferred` to preserve `Deferred` return type. Then the function return type hint would be simply `-> dict[str, str]`.

Am I missing anything?

Reply:
To do this conversion there is still some things that are unclear to me:

* I often see a yield on an object that is not a deferred one. If I convert this to await it will be a runtime error. However if do not convert this to an await, is there potential other code that might override the function in a subclass and makes it async?
* Also how does twisted deals with deferred objects that are never yield? With await these futures might be never executed. In my own project I had deferred object that had never been yielded and there were fine somehow. Without having types everywhere, this seems very dangerous because now I have to check the whole code base if there is any method call that is not deferred.

With just adding types so far it seems relatively safe to not change the behavior too much. However with async I have the feeling, I can potentially introduce many nasty bugs.

Reply:
> I often see a yield on an object that is not a deferred one. If I convert this to await it will be a runtime error. However if do not convert this to an await, is there potential other code that might override the function in a subclass and makes it async?

Thanks, I haven't realized that. Both ways possible - Deferred to non-Deferred and vice-versa. We'll need something like the following in all places where we await something that user may extend.

```
async def maybe_async(c):
    if inspect.isawaitable(c) or isinstance(c, defer.Deferred):
        return await c
    return c
```

> Also how does twisted deals with deferred objects that are never yield? With await these futures might be never executed.

Twisted still executes them. However if we decorate the function with `async_to_deferred` it will still be awaited in all cases and bug will not happen.

Is it possible to define some type that instead of `Generator[defer.Deferred[Any], Any, dict[str, str]]` one would write e.g. `DeferredGenerator[dict[str, str]]`? That would alleviate my concerns.

Reply:
> Thanks, I haven't realized that. Both ways possible - Deferred to non-Deferred and vice-versa. We'll need something like the following in all places where we await something that user may extend.

Twisted already has the helper `defer.maybeDeferred` that does exactly this.

> Is it possible to define some type that instead of Generator[defer.Deferred[Any], Any, dict[str, str]] one would write e.g. DeferredGenerator[dict[str, str]]? That would alleviate my concerns.

It is possible, see https://github.com/buildbot/buildbot/issues/7371#issuecomment-1956564405

Reply:
> Twisted already has the helper defer.maybeDeferred that does exactly this.

Yeah, the only reason why I wrote maybe_async was not to go through Deferred if the argument is coroutine already.

> It is possible, see https://github.com/buildbot/buildbot/issues/7371#issuecomment-1956564405

Cool, I haven't read that comment until now.

Reply:
> Yeah, the only reason why I wrote maybe_async was not to go through Deferred if the argument is coroutine already.

While it would be better, wouldn't it be a bit unsafe until Buildbot require Twisted min version to be 24.7.0 due to [this issue that you patched](https://github.com/twisted/twisted/pull/12239)?

Reply:
https://github.com/twisted/twisted/pull/12239 is about users doing `yield awaitable`. Our plan would be to decorate all functions at API boundaries with `async_to_deferred`. Accordingly, API users would never get a coroutine, but a normal Deferred.

Reply:
@p12tic what are api boundaries actually? All functions without a _?

Reply:
All documented functions. If one cannot find function in the documentation, then it's not an exposed API.

Reply:
@Mic92 I re-read this discussion and it seems that while we covered a lot of topics, what in the end is important is that using `InlineCallbacksType[dict[str, int]]` instead of `Generator[defer.Deferred[Any], Any, dict[str, str]]` would make this PR good to go. The rest of suggestions were either incorrect or could be applied in the future.

InlineCallbacksType would be defined as follows (taken from https://github.com/buildbot/buildbot/issues/7371#issuecomment-1956564405 by @tdesveaux)

```
from typing import Any
from typing import Generator
from typing import TypeVar

T = TypeVar('T')
InlineCallbacksType = Generator[defer.Deferred[Any], Any, T]
```